eng/versioning/pom_file_version_scanner.ps1

```diff
@@ -0,0 +1,370 @@
+            if (!$versionNode.NextSibling.Value.Trim() -like ("{x-version-update;*$($groupId):$($artifactId);"))
+            {
+                $script:FoundError = $true
+                Write-Output "Error: dependency version update tag for groupId=$($groupId), artifactId=$($artifactId) should be <!-- {x-version-update;$($groupId):$($artifactId);current|dependency|external_dependency<select one>} -->"
+            }
+            else
+            {
+                # verify the version tag and version are correct
+                $retVal = Test-Dependency-Tag-And-Version $libHash $extDepHash $versionNode.InnerText $versionNode.NextSibling.Value
+                if ($retVal)
+                {
+                    $script:FoundError = $true
+                    Write-Output "$($retVal)"
+                }
+            }
+        }
+        else 
+        {   
+            $script:FoundError = $true
+            # <!-- {x-version-update;<groupId>:<artifactId>;current} -->
+            Write-Output "Error: Missing dependency version update tag for groupId=$($groupId), artifactId=$($artifactId). The tag should be <!-- {x-version-update;$($groupId):$($artifactId);current|dependency|external_dependency<select one>} -->"
+        }
+    }
+    # Verify every plugin as a group, artifact and version
+    # Verify every dependency as a group, artifact and version
+    # GetElementsByTagName should get all dependencies including dependencies under plugins
+    foreach($pluginNode in $xmlPomFile.GetElementsByTagName("plugin"))
+    {
+        $artifactId = $pluginNode.artifactId
+        $groupId = $pluginNode.groupId
+        # plugins will always have an artifact but may not have a groupId
+        if (!$groupId)
+        {
+            Write-Output "Error: plugin $($artifactId) is missing its groupId tag"
+        }
+        $versionNode = $pluginNode.GetElementsByTagName("version")[0]
+        if ($versionNode.NextSibling -and $versionNode.NextSibling.NodeType -eq "Comment") 
+        {
+            # unfortunately because there are POM exceptions we need to wildcard the group which may be 
+            # something like <area>_groupId
+            if (!$versionNode.NextSibling.Value.Trim() -like ("{x-version-update;*$($groupId):$($artifactId);"))
+            {
+                $script:FoundError = $true
+                Write-Output "Error: plugin version update tag for groupId=$($groupId), artifactId=$($artifactId) should be <!-- {x-version-update;$($groupId):$($artifactId);current|dependency|external_dependency<select one>} -->"
+            }
+            else
+            {
+                # verify the version tag and version are correct
+                $retVal = Test-Dependency-Tag-And-Version $libHash $extDepHash $versionNode.InnerText $versionNode.NextSibling.Value
+                if ($retVal)
+                {
+                    $script:FoundError = $true
+                    Write-Output "$($retVal)"
+                }
```

Comment:
This block of code appears to be common, any chance this could become a function?

Reply:
No, unfortunately not. Write-Output will not work from inside a function (not in the way any sane person would think it would) which is why I'd opted to return the strings. Write-Error would work but the formatting of Write-Error is such where I'm only trying to use that when there are really dire (like malformed version/dependency files). For missing tags or versions I'm intentionally using Write-Output to list everything that's wrong for a given pom file without all the embellishments that Write-Error gives.